File: sdks/go/tests/conformance/collector_api_cli_test.go
package conformance

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/astraive/loxa/sdks/go"
)

func TestCollectorAndCortexClientFamilies(t *testing.T) {
	collectorServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		switch r.URL.Path {
		case "/health":
			w.WriteHeader(http.StatusOK)
			_, _ = w.Write([]byte(`{"status":"ok"}`))
		case "/validate":
			w.WriteHeader(http.StatusOK)
			_, _ = w.Write([]byte(`{"valid":true}`))
		case "/ingest":
			w.WriteHeader(http.StatusOK)
			_, _ = w.Write([]byte(`{"accepted":1}`))
		case "/query":
			w.WriteHeader(http.StatusOK)
			_, _ = w.Write([]byte(`{"events":[]}`))
		case "/tail":
			w.WriteHeader(http.StatusOK)
			_, _ = w.Write([]byte(`{"events":[]}`))
		case "/events":
			w.WriteHeader(http.StatusOK)
			_, _ = w.Write([]byte(`{"deleted":0}`))
		case "/replay":
			w.WriteHeader(http.StatusOK)
			_, _ = w.Write([]byte(`{"replayed":0}`))
		case "/dlq/list":
			w.WriteHeader(http.StatusOK)
			_, _ = w.Write([]byte(`{"entries":[]}`))
		case "/dlq/item":
			w.WriteHeader(http.StatusOK)
			_, _ = w.Write([]byte(`{"entry":null}`))
		case "/dlq/item/replay":
			w.WriteHeader(http.StatusOK)
			_, _ = w.Write([]byte(`{"replayed":1}`))
		case "/keys":
			if r.Method == http.MethodPost {
				w.WriteHeader(http.StatusCreated)
				_, _ = w.Write([]byte(`{"id":"key_123"}`))
			} else {
				w.WriteHeader(http.StatusNotFound)
			}
		case "/keys/key_123":
			if r.Method == http.MethodDelete {
				w.WriteHeader(http.StatusOK)
				_, _ = w.Write([]byte(`{"revoked":true}`))
			} else {
				w.WriteHeader(http.StatusNotFound)
			}
		case "/keys/key_123/rotate":
			w.WriteHeader(http.StatusOK)
			_, _ = w.Write([]byte(`{"rotated":true}`))
		case "/sinks":
			w.WriteHeader(http.StatusOK)
			_, _ = w.Write([]byte(`{"sinks":[]}`))
		case "/sinks/stdout/test":
			w.WriteHeader(http.StatusOK)
			_, _ = w.Write([]byte(`{"status":"healthy"}`))
		case "/policy/validate":
			w.WriteHeader(http.StatusOK)
			_, _ = w.Write([]byte(`{"valid":true,"errors":[]}`))
		case "/schema/check":
			w.WriteHeader(http.StatusOK)
			_, _ = w.Write([]byte(`{"valid":true}`))
		case "/schema/publish":
			w.WriteHeader(http.StatusOK)
			_, _ = w.Write([]byte(`{"published":true}`))
		case "/retention/apply":
			w.WriteHeader(http.StatusOK)
			_, _ = w.Write([]byte(`{"applied":true}`))
		default:
			w.WriteHeader(http.StatusNotFound)
		}
	}))
	defer collectorServer.Close()

	collectorClient := loxa.NewCollectorClient(loxa.CollectorClientConfig{
		Endpoint: collectorServer.URL,
		Client:   collectorServer.Client(),
	})
	if err := collectorClient.Health(context.Background()); err != nil {
		t.Fatalf("collector health: %v", err)
	}
	if _, err := collectorClient.Validate(context.Background(), json.RawMessage(`{"event":"verification"}`)); err != nil {
		t.Fatalf("collector validate: %v", err)
	}
	if _, err := collectorClient.Ingest(context.Background(), []json.RawMessage{json.RawMessage(`{"event":"verification"}`)}); err != nil {
		t.Fatalf("collector ingest: %v", err)
	}
	if _, err := collectorClient.Query(context.Background(), json.RawMessage(`{"query":"select 1"}`)); err != nil {
		t.Fatalf("collector query: %v", err)
	}
	if _, err := collectorClient.Tail(context.Background(), json.RawMessage(`{"limit":1}`)); err != nil {
		t.Fatalf("collector tail: %v", err)
	}
	if _, err := collectorClient.Delete(context.Background(), json.RawMessage(`{"event":"verification"}`)); err != nil {
		t.Fatalf("collector delete: %v", err)
	}
	if _, err := collectorClient.Replay(context.Background(), json.RawMessage(`{"event_ids":["evt_1"]}`)); err != nil {
		t.Fatalf("collector replay: %v", err)
	}
	if _, err := collectorClient.DLQList(context.Background(), json.RawMessage(`{"limit":1}`)); err != nil {
		t.Fatalf("collector dlq list: %v", err)
	}
	if _, err := collectorClient.DLQRead(context.Background(), "item"); err != nil {
		t.Fatalf("collector dlq read: %v", err)
	}
	if _, err := collectorClient.DLQReplay(context.Background(), "item"); err != nil {
		t.Fatalf("collector dlq replay: %v", err)
	}
	if _, err := collectorClient.KeysCreate(context.Background(), json.RawMessage(`{"name":"verification"}`)); err != nil {
		t.Fatalf("collector keys create: %v", err)
	}
	if _, err := collectorClient.KeysRevoke(context.Background(), "key_123"); err != nil {
		t.Fatalf("collector keys revoke: %v", err)
	}
	if _, err := collectorClient.KeysRotate(context.Background(), "key_123"); err != nil {
		t.Fatalf("collector keys rotate: %v", err)
	}
	if _, err := collectorClient.SinksList(context.Background()); err != nil {
		t.Fatalf("collector sinks list: %v", err)
	}
	if _, err := collectorClient.SinksTest(context.Background(), "stdout"); err != nil {
		t.Fatalf("collector sinks test: %v", err)
	}
	if _, err := collectorClient.PolicyValidate(context.Background(), json.RawMessage(`{"sample_rate":1.0}`)); err != nil {
		t.Fatalf("collector policy validate: %v", err)
	}
	if _, err := collectorClient.SchemaCheck(context.Background(), json.RawMessage(`{"event":"verification"}`)); err != nil {
		t.Fatalf("collector schema check: %v", err)
	}
	if _, err := collectorClient.SchemaPublish(context.Background(), json.RawMessage(`{"schema":"v1"}`)); err != nil {
		t.Fatalf("collector schema publish: %v", err)
	}
	if _, err := collectorClient.RetentionApply(context.Background(), json.RawMessage(`{"days":7}`)); err != nil {
		t.Fatalf("collector retention apply: %v", err)
	}

	cortexServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		switch r.URL.Path {
		case "/healthz", "/readyz":
			w.WriteHeader(http.StatusOK)
			_, _ = w.Write([]byte(`{"status":"ok","ready":true}`))
		case "/reconstruct":
			w.WriteHeader(http.StatusOK)
			_, _ = w.Write([]byte(`{"incident_id":"inc_123","timestamp":"2026-05-23T00:00:00Z","confidence":0.9,"related_services":["checkout"],"related_events":[],"causal_chain":[],"similar_past_incidents":[],"suggested_remediations":[],"symptoms":[],"explain":"ok"}`))
		case "/graph/service/checkout":
			w.WriteHeader(http.StatusOK)
			_, _ = w.Write([]byte(`{"nodes":[{"id":"checkout"}],"edges":[]}`))
		case "/graph/incident/inc_123":
			w.WriteHeader(http.StatusOK)
			_, _ = w.Write([]byte(`{"nodes":[{"id":"checkout"}],"edges":[]}`))
		case "/feedback/remediation", "/feedback/incident", "/events/batch":
			w.WriteHeader(http.StatusOK)
			_, _ = w.Write([]byte(`{"ok":true}`))
		default:
			w.WriteHeader(http.StatusNotFound)
		}
	}))
	defer cortexServer.Close()

	cortexClient := loxa.NewCortexClient(cortexServer.URL)
	if !cortexClient.Health(context.Background()) {
		t.Fatalf("expected cortex health")
	}
	if !cortexClient.Ready(context.Background()) {
		t.Fatalf("expected cortex ready")
	}
	if _, err := cortexClient.Reconstruct(context.Background(), "inc_123", "fast"); err != nil {
		t.Fatalf("cortex reconstruct: %v", err)
	}
	if _, err := cortexClient.ServiceGraph(context.Background(), "checkout", 1); err != nil {
		t.Fatalf("cortex service graph: %v", err)
	}
	if _, err := cortexClient.IncidentGraph(context.Background(), "inc_123", 1); err != nil {
		t.Fatalf("cortex incident graph: %v", err)
	}
	if err := cortexClient.RecordRemediation(context.Background(), &loxa.Remediation{IncidentID: "inc_123", Action: "restart"}); err != nil {
		t.Fatalf("cortex remediation: %v", err)
	}
	if err := cortexClient.RecordFeedback(context.Background(), &loxa.RemediationFeedback{RemediationID: "rem_123", IncidentID: "inc_123", Outcome: "success"}); err != nil {
		t.Fatalf("cortex feedback: %v", err)
	}
	if err := cortexClient.IngestBatch(context.Background(), []map[string]any{{"event": "verification"}}); err != nil {
		t.Fatalf("cortex ingest batch: %v", err)
	}
	if err := loxa.ValidateIncidentContext(&loxa.IncidentContext{}); err == nil {
		t.Fatalf("expected incident context validation error")
	}
}

File: sdks/js/tests/collector_api_cli.test.ts
import { afterEach, it } from 'node:test';
import assert from 'node:assert/strict';
import http from 'node:http';
import * as loza from '../src/index.ts';

afterEach(() => {
  loza.reset();
});

it('covers collector and cortex client families', async () => {
  const collectorServer = http.createServer((req, res) => {
    const url = new URL(req.url || '/', 'http://localhost');
    const path = url.pathname;
    const method = req.method || 'GET';
    const writeJson = (status: number, obj: any) => {
      res.writeHead(status, { 'Content-Type': 'application/json' });
      res.end(JSON.stringify(obj));
    };
    if (path === '/health' || path === '/ready') return writeJson(200, { status: 'ok' });
    if (path === '/version') return writeJson(200, { version: '0.2.0', ingest_api_version: 'v1', schema_version: 'v1', event_version: 'v1' });
    if (path === '/validate') return writeJson(200, { valid: true });
    if (path === '/events' && method === 'POST') return writeJson(200, { accepted: 1, rejected: 0, invalid: 0 });
    if (path === '/query') return writeJson(200, { rows: [] });
    if (path === '/tail' && method === 'GET') return writeJson(200, { events: [] });
    if (path === '/events' && method === 'DELETE') return writeJson(200, { deleted: 1 });
    if (path === '/replay') return writeJson(202, { replayed: 1 });
    if (path === '/dlq') return writeJson(200, { events: [] });
    if (path.match(/^\/dlq\/[^/]+$/) && !path.endsWith('/replay')) return writeJson(200, { entry: {} });
    if (path.match(/^\/dlq\/[^/]+\/replay$/)) return writeJson(200, { replayed: 1 });
    if (path === '/keys' && method === 'POST') return writeJson(201, { id: 'k_1' });
    if (path.match(/^\/keys\/[^/]+$/) && method === 'DELETE') return writeJson(200, { revoked: true });
    if (path.match(/^\/keys\/[^/]+\/rotate$/)) return writeJson(200, { rotated: true });
    if (path === '/sinks') return writeJson(200, { sinks: [] });
    if (path.match(/^\/sinks\/[^/]+\/test$/)) return writeJson(200, { status: 'healthy' });
    if (path === '/policy/validate') return writeJson(200, { valid: true, errors: [] });
    if (path === '/schema/check') return writeJson(200, { valid: true });
    if (path === '/schema/publish') return writeJson(200, { published: true });
    if (path === '/retention/apply') return writeJson(200, { applied: true });
    writeJson(404, { error: 'not_found' });
  });
  await new Promise<void>((resolve) => collectorServer.listen(0, resolve));
  const collectorPort = (collectorServer.address() as any).port;
  const collector = new loza.CollectorClient({ url: `http://127.0.0.1:${collectorPort}`, enableCompression: false });

  assert.equal(await collector.health(), true);
  assert.equal(await collector.ready(), true);
  assert.equal((await collector.version()).version, '0.2.0');
  assert.equal((await collector.validate({ event: 'catalog.validate' })).valid, true);
  await collector.ingest([{ event_id: 'evt_1', event: 'catalog.ingest' }]);
  await collector.query({ query: 'select 1' });
  await collector.tail({ limit: 1 });
  await collector.delete({ event: 'catalog.delete' });
  await collector.replay({ event_ids: ['evt_1'] });
  await collector.dlqList({ limit: 1 });
  await collector.dlqRead('dlq_1');
  await collector.dlqReplay('dlq_1');
  await collector.keysCreate('catalog');
  await collector.keysRevoke('key_1');
  await collector.keysRotate('key_1');
  await collector.sinksList();
  await collector.sinksTest('stdout');
  assert.equal((await collector.policyValidate({ sample_rate: 1.0 })).valid, true);
  assert.equal((await collector.schemaCheck({ event: 'verification' })).valid, true);
  assert.equal((await collector.schemaPublish({ schema: 'v1' })).published, true);
  assert.equal((await collector.retentionApply({ days: 7 })).applied, true);
  await new Promise<void>((resolve) => collectorServer.close(() => resolve()));

  const cortexServer = http.createServer((req, res) => {
    const path = req.url || '/';
    const writeJson = (status: number, obj: any) => {
      res.writeHead(status, { 'Content-Type': 'application/json' });
      res.end(JSON.stringify(obj));
    };
    if (path === '/healthz' || path === '/readyz') return writeJson(200, { status: 'ok', ready: true });
    if (path.startsWith('/graph/')) return writeJson(200, { nodes: [{ id: 'checkout' }], edges: [] });
    writeJson(200, {
      incident_id: 'inc_123',
      timestamp: '2026-05-23T00:00:00Z',
      confidence: 0.9,
      related_services: ['checkout'],
      related_events: [],
      causal_chain: [],
      similar_past_incidents: [],
      suggested_remediations: [],
      symptoms: [],
      explain: 'ok',
    });
  });
  await new Promise<void>((resolve) => cortexServer.listen(0, resolve));
  const cortexPort = (cortexServer.address() as any).port;
  const cortex = new loza.CortexClient({ url: `http://127.0.0.1:${cortexPort}` });

  assert.equal(await cortex.health(), true);
  assert.equal(await cortex.ready(), true);
  assert.equal((await cortex.reconstruct('inc_123')).incident_id, 'inc_123');
  assert.equal((await cortex.similarIncidents('inc_123')).length, 0);
  assert.equal((await cortex.serviceGraph('checkout')).nodes[0].id, 'checkout');
  assert.equal((await cortex.incidentGraph('inc_123')).nodes[0].id, 'checkout');
  await cortex.recordRemediation({ incident_id: 'inc_123', action: 'restart' });
  await cortex.recordFeedback({ incident_id: 'inc_123', outcome: 'success' });
  await cortex.ingestBatch([{ event: 'catalog' }]);
  await cortex.ingestJsonl([{ event: 'catalog' }]);
  await new Promise<void>((resolve) => cortexServer.close(() => resolve()));
});
